protocol/auditlog_test.go

```diff
@@ -0,0 +1,200 @@
+package protocol
+
+import (
+	"github.com/coniks-sys/coniks-go/crypto"
+	"testing"
+)
+
+func TestInsertEmptyHistory(t *testing.T) {
+	// create basic test directory and audit log with 1 STR
+	_, _, _ = NewTestAuditLog(t, 0)
+}
+
+func TestUpdateHistory(t *testing.T) {
+	// create basic test directory and audit log with 1 STR
+	d, aud, hist := NewTestAuditLog(t, 0)
+
+	// update the directory so we can update the audit log
+	dirInitHash := computeInitSTRHash(hist[0])
+	d.Update()
+	err := aud.Update(dirInitHash, d.LatestSTR())
+
+	if err != nil {
+		t.Fatal("Error updating the server history")
+	}
+}
+
+func TestInsertPriorHistory(t *testing.T) {
+	// create basic test directory and audit log with 11 STRs
+	_, _, _ = NewTestAuditLog(t, 10)
+}
+
+func TestInsertExistingHistory(t *testing.T) {
+	// create basic test directory and audit log with 1 STR
+	_, aud, hist := NewTestAuditLog(t, 0)
+
+	// let's make sure that we can't re-insert a new server
+	// history into our log
+	err := aud.Insert("test-server", nil, hist)
+	if err != ErrAuditLog {
+		t.Fatal("Expected an ErrAuditLog when inserting an existing server history")
+	}
+}
+
+func TestUpdateUnknownHistory(t *testing.T) {
+	// create basic test directory and audit log with 1 STR
+	d, aud, _ := NewTestAuditLog(t, 0)
+
+	// let's make sure that we can't update a history for an unknown
+	// directory in our log
+	var unknown [crypto.HashSizeByte]byte
+	err := aud.Update(unknown, d.LatestSTR())
+	if err != ErrAuditLog {
+		t.Fatal("Expected an ErrAuditLog when updating an unknown server history")
+	}
+}
+
+func TestUpdateBadNewSTR(t *testing.T) {
+	// create basic test directory and audit log with 11 STRs
+	d, aud, hist := NewTestAuditLog(t, 10)
+
+	// compute the hash of the initial STR for later lookups
+	dirInitHash := computeInitSTRHash(hist[0])
+
+	// update the directory a few more times and then try
+	// to update
+	d.Update()
+	d.Update()
+
+	err := aud.Update(dirInitHash, d.LatestSTR())
+	if err != CheckBadSTR {
+		t.Fatal("Expected a CheckBadSTR when attempting update a server history with a bad STR")
+	}
+}
+
+func TestGetLatestObservedSTR(t *testing.T) {
+	// create basic test directory and audit log with 1 STR
+	d, aud, hist := NewTestAuditLog(t, 0)
+
+	// compute the hash of the initial STR for later lookups
+	dirInitHash := computeInitSTRHash(hist[0])
+
+	res, err := aud.GetObservedSTRs(&AuditingRequest{
+		DirInitSTRHash: dirInitHash,
+		StartEpoch:     uint64(d.LatestSTR().Epoch),
+		EndEpoch:       uint64(d.LatestSTR().Epoch)})
+	if err != ReqSuccess {
+		t.Fatal("Unable to get latest observed STR")
+	}
+
+	obs := res.DirectoryResponse.(*STRHistoryRange)
+	if len(obs.STR) == 0 {
+		t.Fatal("Expect returned STR to be not nil")
+	}
+	if obs.STR[0].Epoch != d.LatestSTR().Epoch {
+		t.Fatal("Unexpected epoch for returned latest STR")
+	}
```

Comment:
I think what you're looking for is in `TestGetEntireSTRHist()`. Since the latestSTR is always inserted into the `snapshots` on each `aud.Update()`, the problem Arlo mentioned should be taken care of now.